[SPARK-14978][PySpark] PySpark TrainValidationSplitModel should support validationMetrics

taku-k opened the pull request:
## What changes were proposed in this pull request?

This pull request includes supporting validationMetrics for TrainValidationSplitModel with Python and test for it.
## How was this patch tested?

test in `python/pyspark/ml/tests.py`

taku-k committed:
support validationMetrics in PySpark TrainValidationSplitModel

AmplabJenkins commented:
Can one of the admins verify this patch?

vectorijk commented on python/pyspark/ml/tuning.py:
```diff
@@ -613,7 +615,9 @@ def copy(self, extra=None):
-        return TrainValidationSplitModel(self.bestModel.copy(extra))
+        bestModel = self.bestModel.copy(extra)
+        validationMetrics = self.validationMetrics
+        return TrainValidationSplitModel(bestModel, validationMetrics)
```
It would be great if you could add test case for `TrainValidationSplitModel` copy method.

taku-k replied:
@vectorijk Thank you for your comments. I'm sure. I'll add `test_copy` method.

vectorijk commented on python/pyspark/ml/tests.py:
```diff
@@ -616,6 +622,7 @@ def test_save_load(self):
         tvsModel.save(tvsModelPath)
         loadedModel = TrainValidationSplitModel.load(tvsModelPath)
         self.assertEqual(loadedModel.bestModel.uid, tvsModel.bestModel.uid)
+        self.assertEqual(len(loadedModel.validationMetrics), len(tvsModel.validationMetrics))
```
It would be better to compare each `validationMetric` in the list not only the length of `validationMetrics`   cc @jkbradley

taku-k replied:
I agree!

vectorijk commented on python/pyspark/ml/tests.py:
```diff
@@ -586,10 +589,13 @@ def test_fit_maximize_metric(self):
         self.assertEqual(0.0, bestModel.getOrDefault('inducedError'),
                          "Best model should have zero induced error")
         self.assertEqual(1.0, bestModelMetric, "Best model has R-squared of 1")
+        self.assertEqual(len(grid), len(validationMetrics),
```
same as here I suppose.

taku-k replied:
Here, it is checked that a `validationMetrics` inside `TrainValidationSplitModel` is the same size as  given grid parameters.
So, I think comparing each validationMetric in the list is not needed.

MLnick replied:
It may be overkill but it should be easy to also test that the best value in `validationMetrics` equals `bestModelMetric`

sethah replied:
Let's make this test line up with the minimize version by checking that the best model metric is the max of validation metrics.

taku-k committed:
add more tests

add test_copy method for testing copy
add test case for comparing each validationMetric

taku-k commented:
cc @feynmanliang @jkbradley @mengxr

MLnick commented:
@taku-k please rebase your PR to latest master and I'll review. You'll need to adjust the tests related to save/load since [SPARK-13786](https://issues.apache.org/jira/browse/SPARK-13786) was reverted.

taku-k commented:
@MLnick Thank you. I rebased to latest master and resolved conflicts. I would like you to review my Pull/Req.

MLnick commented on python/pyspark/ml/tests.py:
```diff
@@ -645,6 +651,37 @@ def test_save_load(self):
+        iee = InducedErrorEstimator()
+        evaluator = RegressionEvaluator(metricName="r2")
+
+        grid = (ParamGridBuilder()
```
is the extra `(` and `)` necessary here?

taku-k replied:
I need additional parenthesis to call chained method with indentation style.

holdenk replied:
There is also the multi-line '\' syntax

MLnick replied:
@holdenk @taku-k I think we should prefer `\` syntax

MLnick commented on python/pyspark/ml/tests.py:
```diff
@@ -645,6 +651,37 @@ def test_save_load(self):
+        self.assertEqual(tvs.getEstimator().uid, tvsCopied.getEstimator().uid,
+                         "Copied TrainValidationSplit has the same uid of Estimator")
+
+        self.assertEqual(len(tvsModel.validationMetrics),
```
Since we're testing the copy method here, we should add tests for `bestModel` also

MLnick commented:
ok to test

SparkQA commented:
**[Test build #58063 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58063/consoleFull)** for PR 12767 at commit [`02e9642`](https://github.com/apache/spark/commit/02e9642094a410f3f365da70e4cbbfbd4692cc89).

SparkQA commented:
**[Test build #58063 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58063/consoleFull)** for PR 12767 at commit [`02e9642`](https://github.com/apache/spark/commit/02e9642094a410f3f365da70e4cbbfbd4692cc89).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58063/
Test FAILed.

taku-k committed:
improve test and remove redundant blank line

SparkQA commented:
**[Test build #58075 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58075/consoleFull)** for PR 12767 at commit [`88c40d4`](https://github.com/apache/spark/commit/88c40d426c89160be4dc12fae3fa13e664fd5027).

SparkQA commented:
**[Test build #58075 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58075/consoleFull)** for PR 12767 at commit [`88c40d4`](https://github.com/apache/spark/commit/88c40d426c89160be4dc12fae3fa13e664fd5027).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58075/
Test PASSed.

MLnick commented:
cc @holdenk @sethah

sethah commented on python/pyspark/ml/tests.py:
```diff
@@ -590,10 +590,14 @@ def test_fit_minimize_metric(self):
         self.assertEqual(0.0, bestModelMetric, "Best model has RMSE of 0")
+        self.assertEqual(len(grid), len(validationMetrics),
+                         "validationMetrics has the same size of grid parameter")
+        self.assertIn(bestModelMetric, validationMetrics)
```
`self.assertEqual(bestModelMetric, max(validationMetrics))`? It's more explicit IMO.

MLnick replied:
Just a note on this. In general, this is actually not true - the `bestModelMetric` is the metric achieved on the full dataset. Whereas the `max(validationMetrics)` is the metric that the best model achieved on the validation set within `TrainValidationSplit`. It happens to be the same here as we induce a 0 error.

The test is valid though, I thought I would just mention it.

We could also add the same test to `CrossValidator` tests (as again it happens to work for the 0 induced error).

holdenk commented on python/pyspark/ml/tuning.py:
```diff
@@ -431,7 +433,9 @@ def copy(self, extra=None):
             extra = dict()
-        return TrainValidationSplitModel(self.bestModel.copy(extra))
+        bestModel = self.bestModel.copy(extra)
+        validationMetrics = self.validationMetrics
```
It might be good to explicitly also copy the validation metrics rather than just the reference?

taku-k committed:
change test and indentiation style

SparkQA commented:
**[Test build #58599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58599/consoleFull)** for PR 12767 at commit [`b5879b7`](https://github.com/apache/spark/commit/b5879b719291b548f4a2ace00ba155b23b61c050).

SparkQA commented:
**[Test build #58599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58599/consoleFull)** for PR 12767 at commit [`b5879b7`](https://github.com/apache/spark/commit/b5879b719291b548f4a2ace00ba155b23b61c050).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58599/
Test FAILed.

SparkQA commented:
**[Test build #58600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58600/consoleFull)** for PR 12767 at commit [`f08d064`](https://github.com/apache/spark/commit/f08d0649f01bd32720e5039a0207fad12c64a224).

SparkQA commented:
**[Test build #58600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58600/consoleFull)** for PR 12767 at commit [`f08d064`](https://github.com/apache/spark/commit/f08d0649f01bd32720e5039a0207fad12c64a224).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58600/
Test FAILed.

taku-k committed:
use SparkSession instead of SQLContext

SparkQA commented:
**[Test build #58615 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58615/consoleFull)** for PR 12767 at commit [`2948eb0`](https://github.com/apache/spark/commit/2948eb0ab45ea03324075d6a351c18deca7e411c).

SparkQA commented:
**[Test build #58615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58615/consoleFull)** for PR 12767 at commit [`2948eb0`](https://github.com/apache/spark/commit/2948eb0ab45ea03324075d6a351c18deca7e411c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58615/
Test PASSed.

MLnick commented:
This LGTM, cc @sethah @holdenk for a final pass?

holdenk commented on python/pyspark/ml/tests.py:
```diff
@@ -594,17 +594,21 @@ def test_fit_minimize_metric(self):
-                .addGrid(iee.inducedError, [100.0, 0.0, 10000.0])
-                .build())
+        grid = ParamGridBuilder() \
+            .addGrid(iee.inducedError, [100.0, 0.0, 10000.0])\
```
super minor nit here but I think generally we have ' \' rather than '\' for multi-line (as you've done elsewhere in this PR).

taku-k replied:
Thanks!

SparkQA commented:
**[Test build #58632 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58632/consoleFull)** for PR 12767 at commit [`8c28f3c`](https://github.com/apache/spark/commit/8c28f3ccc3dd315a0b01b5be8ff40fb8f8ac1782).

holdenk commented on python/pyspark/ml/tuning.py:
```diff
@@ -445,7 +447,9 @@ def copy(self, extra=None):
             extra = dict()
-        return TrainValidationSplitModel(self.bestModel.copy(extra))
+        bestModel = self.bestModel.copy(extra)
+        validationMetrics = list(self.validationMetrics)
```
So this is a shallow copy rather than a deep copy - its probably OK since we aren't explicit in the docstring but maybe we should be explicit in the docstring that we make a shallow copy of the validation metrics since we are explicit about the other parts of the copy in the docstring?

SparkQA commented:
**[Test build #58632 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58632/consoleFull)** for PR 12767 at commit [`8c28f3c`](https://github.com/apache/spark/commit/8c28f3ccc3dd315a0b01b5be8ff40fb8f8ac1782).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58632/
Test PASSed.

taku-k committed:
update docstring of TrainValidationSplitModel.copy

SparkQA commented:
**[Test build #58634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58634/consoleFull)** for PR 12767 at commit [`90e1092`](https://github.com/apache/spark/commit/90e10927983f58203643e71c61fe4ac4fad902a3).

SparkQA commented:
**[Test build #58634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58634/consoleFull)** for PR 12767 at commit [`90e1092`](https://github.com/apache/spark/commit/90e10927983f58203643e71c61fe4ac4fad902a3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58634/
Test PASSed.

sethah commented on python/pyspark/ml/tests.py:
```diff
@@ -594,17 +594,21 @@ def test_fit_minimize_metric(self):
         self.assertEqual(0.0, bestModelMetric, "Best model has RMSE of 0")
+        self.assertEqual(len(grid), len(validationMetrics),
+                         "validationMetrics has the same size of grid parameter")
+        self.assertEqual(bestModelMetric, min(validationMetrics))
```
With @MLnick's previous comment in mind, my preference would be to instead check:

`self.assertEqual(0.0, min(validationMetrics))`

That way we don't put in a check that is wrong in general, but correct here. I don't feel too strongly about it, but since it's an easy fix and we might save some future developer a bit of confusion then I think it's a good idea. Same in the maximize version.

sethah commented:
LGTM pending one small issue.

taku-k committed:
change to suitable tests

SparkQA commented:
**[Test build #58698 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58698/consoleFull)** for PR 12767 at commit [`a3adef5`](https://github.com/apache/spark/commit/a3adef557471abdf1326c79a89d4c67cbbf53605).

SparkQA commented:
**[Test build #58698 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58698/consoleFull)** for PR 12767 at commit [`a3adef5`](https://github.com/apache/spark/commit/a3adef557471abdf1326c79a89d4c67cbbf53605).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58698/
Test PASSed.

MLnick commented:
Merged to master/branch-2.0. Thanks @vectorijk @holdenk @sethah for the review!

MLnick commented:
@taku-k please provide your JIRA username so I can assign the ticket to you.

taku-k commented:
@MLnick Thanks! My JIRA username is `taku-k`. It's the same as the github username.
